scripts/get-badges.ts

```diff
@@ -1,69 +1,135 @@
 
-Usage: node ${scriptName} PROJECT_ID VENUE_ID CREDENTIAL_PATH
+import { WithId, withId } from "../src/utils/id";
+import { formatSecondsAsDuration } from "../src/utils/time";
 
-Example: node ${scriptName} co-reality-map myawesomevenue prodAccountKey.json
----------------------------------------------------------
-`;
+import { initFirebaseAdminApp, makeScriptUsage } from "./lib/helpers";
 
-  console.log(helpText);
-  process.exit(1);
-};
+const usage = makeScriptUsage({
+  description:
+    "Retrieve 'badge' details (in CSV format) of users who entered the specified venue(s), and how long they spent in each.",
+  usageParams: "PROJECT_ID VENUE_IDS [CREDENTIAL_PATH]",
+  exampleParams:
+    "co-reality-map venueId,venueId2,venueIdN [theMatchingAccountServiceKey.json]",
+});
 
-const [projectId, venueId] = process.argv.slice(2);
-if (!projectId || !venueId) {
+const [projectId, venueIds, credentialPath] = process.argv.slice(2);
+
+// Note: no need to check credentialPath here as initFirebaseAdmin defaults it when undefined
+if (!projectId || !venueIds) {
   usage();
 }
 
-const app = initFirebaseAdminApp(projectId, {
-  appName: projectId,
+const venueIdsArray = venueIds.split(",");
+
+// Note: if we ever need to handle this, we can split our firestore query into 'chunks', each with 10 items per array-contains-any
+if (venueIdsArray.length > 10) {
+  console.error(
+    "Error: This script can only handle up to 10 venueIds at once at the moment."
+  );
+  console.error("  venueIdsArray.length :", venueIdsArray.length);
+  process.exit(1);
+}
+
+initFirebaseAdminApp(projectId, {
+  credentialPath: credentialPath
+    ? resolve(__dirname, credentialPath)
+    : undefined,
 });
 
+interface UsersWithVisitsResult {
+  user: WithId<User>;
+  visits: WithId<UserVisit>[];
+}
+
 (async () => {
-  const allUsers: admin.auth.UserRecord[] = [];
-  let nextPageToken: string | undefined;
-  const { users, pageToken } = await app.auth().listUsers(1000);
+  // TODO: extract this as a generic helper function?
+  const usersWithVisits = await Promise.all(
+    await admin
+      .firestore()
+      .collection("users")
+      .where("enteredVenueIds", "array-contains-any", venueIdsArray)
+      .get()
+      .then((usersSnapshot) =>
+        usersSnapshot.docs.map(async (userDoc) => {
+          const user = withId(userDoc.data() as User, userDoc.id);
 
-  allUsers.push(...users);
-  nextPageToken = pageToken;
+          const visits = await userDoc.ref
+            .collection("visits")
+            .get()
+            .then((visitsSnapshot) =>
+              visitsSnapshot.docs.map((visitDoc) =>
+                withId(visitDoc.data() as UserVisit, visitDoc.id)
+              )
+            );
 
-  while (nextPageToken) {
-    const { users, pageToken } = await app
-      .auth()
-      .listUsers(1000, nextPageToken);
-    allUsers.push(...users);
-    nextPageToken = pageToken;
-  }
+          const userWithVisits: UsersWithVisitsResult = { user, visits };
```

Comment:
I believe I chose to do it to explicitly add the type to it here, rather than needing to do so at the ‘entry point’ of this bit of code, or ‘force’ the cast with ‘as’ (which I believe would overwrite it even if it was inaccurate).

From memory, if I don’t add the type, then later on in type hints/etc we see the full ‘object of types’ (as it can infer the correct types), rather than the much simpler ‘grouping type name’, which takes longer/more cognitive effort to understand/what it means/is referring to/etc.

It might be fine to add it at the ‘entry point’ rather than here, I would have to look deeper at my code again.

Reply:
Fixed in https://github.com/sparkletown/sparkle/pull/1065/commits/0837f02086f5c879eae48fe479231622d1e14dd5